packages/common/src/hooks/useImageSize2.ts

```diff
@@ -0,0 +1,134 @@
+import { useState, useEffect, useCallback } from 'react'
+
+import { SquareSizes, Track } from '~/models'
+import { Maybe } from '~/utils/typeUtils'
+
+// Global image cache
+const IMAGE_CACHE = new Set<string>()
+
+/**
+ * Fetches an image from the given artwork object managing sizes and using fallback mirrors if necessary.
+ *  - If a larger image has already been fetched and is in the case, use it instead
+ *  - If a smaller image has already been fetched and is in the cache, use it while fetching the larger one
+ *  - If fetching a given image fails, substitute out mirrors and try again
+ * @param artwork - The artwork object to fetch from
+ * @param targetSize - The desired size of the image
+ * @param defaultImage - The fallback image to use if no image is found in the `artwork` object
+ * @returns The url of the image, or undefined if the image is not available
+ */
+export const useImageSize2 = ({
+  artwork,
+  targetSize,
+  defaultImage
+}: {
+  artwork: Maybe<Track['artwork']>
+  targetSize: SquareSizes
+  defaultImage: string
+}) => {
+  const [imageUrl, setImageUrl] = useState<string | null>(null)
+
+  const preloadImage = (url: string) => {
+    return new Promise((resolve, reject) => {
+      const img = new Image()
+      img.src = url
+      img.onload = () => resolve(url)
+      img.onerror = () => reject(url)
+    })
+  }
+
+  const fetchWithFallback = useCallback(
+    async (url: string) => {
+      const mirrors = [...(artwork?.mirrors ?? [])]
+      let currentUrl = url
+
+      while (mirrors.length > 0) {
+        try {
+          await preloadImage(currentUrl)
+          return currentUrl
+        } catch {
+          const nextMirror = mirrors.shift()
+          if (!nextMirror) throw new Error('No mirror found')
+          const currentHost = new URL(currentUrl).hostname
+          const nextHost = new URL(nextMirror).hostname
+          const newUrl = currentUrl.replace(currentHost, nextHost)
+          currentUrl = newUrl
+        }
+      }
+
+      throw new Error('Failed to fetch image from all mirrors')
+    },
+    [artwork?.mirrors]
```

Comment:
I'm curious if this ends up throwing any warnings. React doesn't like it when the length of this array changes between renders, and I don't know if it considers `[undefined]` to be a zero-length array...
If so, might be good to make sure it defaults to a defined value.

Reply:
`[undefined].length` = 1

a defined default could change every render unless there's a global default, but that seems kinda silly? i think this is totally safe & chat.com agrees!